client/js/src/client.ts

```diff
@@ -309,46 +427,430 @@ export async function client(
-								endpoint,
+								endpoint: _endpoint,
 								fn_index
 							});
 							websocket.close();
 						} else if (type === "generating") {
 							fire_event({
 								type: "status",
+								time: new Date(),
 								...status,
-								status: status?.status!,
+								stage: status?.stage!,
 								queue: true,
-								endpoint,
+								endpoint: _endpoint,
 								fn_index
 							});
 						}
 						if (data) {
 							fire_event({
 								type: "data",
+								time: new Date(),
 								data: data.data,
-								endpoint,
+								endpoint: _endpoint,
 								fn_index
 							});
-							res({ data: data.data });
 						}
 					};
 				}
 			});
+
+			function fire_event<K extends EventType>(event: Event<K>) {
+				const narrowed_listener_map: ListenerMap<K> = listener_map;
+				let listeners = narrowed_listener_map[event.type] || [];
+				listeners?.forEach((l) => l(event));
+			}
+
+			function on<K extends EventType>(
+				eventType: K,
+				listener: EventListener<K>
+			) {
+				const narrowed_listener_map: ListenerMap<K> = listener_map;
+				let listeners = narrowed_listener_map[eventType] || [];
+				narrowed_listener_map[eventType] = listeners;
+				listeners?.push(listener);
+
+				return { on, off, cancel };
+			}
+
+			function off<K extends EventType>(
+				eventType: K,
+				listener: EventListener<K>
+			) {
+				const narrowed_listener_map: ListenerMap<K> = listener_map;
+				let listeners = narrowed_listener_map[eventType] || [];
+				listeners = listeners?.filter((l) => l !== listener);
+				narrowed_listener_map[eventType] = listeners;
+
+				return { on, off, cancel };
+			}
+
+			async function cancel() {
+				fire_event({
+					type: "status",
+					endpoint: _endpoint,
+					fn_index: fn_index,
+					stage: "complete",
+					queue: false,
+					time: new Date()
+				});
+
+				try {
+					await fetch(`${http_protocol}//${host + config.path}/reset`, {
+						method: "POST",
+						body: JSON.stringify(session_hash)
+					});
+				} catch (e) {}
+
+				websocket.close();
+			}
+
+			return {
+				on,
+				off,
+				cancel
+			};
 		}
 
-		/**
-		 * Run a prediction.
-		 * @param endpoint - The prediction endpoint to use.
-		 * @param status_callback - A function that is called with the current status of the prediction immediately and every time it updates.
-		 * @return Returns the data for the prediction or an error message.
-		 */
-		function predict(endpoint: string, payload: Payload) {
-			return make_predict(endpoint, payload);
+		async function view_api(
+			config?: Config
+		): Promise<ApiInfo<JsApiData> | [{ error: string }, 500]> {
+			if (api) return api;
+
+			const headers: {
+				Authorization?: string;
+				"Content-Type": "application/json";
+			} = { "Content-Type": "application/json" };
+			if (hf_token) {
+				headers.Authorization = `Bearer ${hf_token}`;
+			}
+			try {
+				let response: Response;
+				// @ts-ignore
+				if (semiver(config.version || "2.0.0", "3.28.3") < 0) {
```

Comment:
Need to update this since the latest changes in @freddyaboulton's PR. Version 29.0.0? or will it be 29.0.1 when the next version is released?

Reply:
For 3.29.0 or older, we should use the space fetcher. Starting from the new release (probably 3.30.0), we can use the /info route.